crates/base-db/src/input.rs

```diff
@@ -657,10 +766,34 @@ impl ops::Index<CrateId> for CrateGraph {
     }
 }
 
+struct ExistingDepsError(Vec<Dependency>);
+
 impl CrateData {
-    fn add_dep(&mut self, dep: Dependency) {
+    /// Add a dependency to `self` without checking if the dependency
+    // is existent among `self.dependencies`.
+    fn add_dep_unchecked(&mut self, dep: Dependency) {
         self.dependencies.push(dep)
     }
+
+    /// Add `deps` to `self` if the dependency is not already listed.
+    /// Finally returning an `Err` propagating the dependencies it couldn't add.
+    fn add_dep(&mut self, deps: Vec<Dependency>) -> Result<(), ExistingDepsError> {
```

Comment:
- Previously the same named function had been adding dependencies to self without checking if the dependency existed. That  function is now called `add_dep_unchecked`. 
- add_dep is not optimal for what it aims to do : I still do not know how well iterators optimize themselves so I just added the most naive linear search code hoping that you could enlighten me on how much/less I should be putting my trust on the compiler.
- Why add_dep return a `Result` : I don't know but I felt like we need to know if a dependency cannot be added.

Reply:
We can remove this again after my proposed changes, but as a note, you overcomplicated your implementation here. Given the invariant, if you insert a dep there should only ever be one duplicate at most as you are preventing duplicates from entering the deps here in the first place. So instead of a vec return value in the err case, a single element would've sufficed.